x/stake/endblock.go

```diff
@@ -252,3 +277,74 @@ func handleMatureUnbondingDelegations(k keeper.Keeper, ctx sdk.Context) ([]types
 
 	return completed, events
 }
+
+const (
+	maxProcessedRefundCount  = 10
+	maxProcessedRefundFailed = 200
+)
+
+func handleRefundStake(ctx sdk.Context, sideChainPrefix []byte, k keeper.Keeper) sdk.Events {
+	sideChainCtx := ctx.WithSideChainKeyPrefix(sideChainPrefix)
+	iterator := k.IteratorAllDelegations(sideChainCtx)
+	defer iterator.Close()
+	var refundEvents sdk.Events
+	succeedCount := 0
+	failedCount := 0
+	boundDenom := k.BondDenom(sideChainCtx)
+	bscSideChainId := k.ScKeeper.BscSideChainId(ctx)
+
+	for ; iterator.Valid(); iterator.Next() {
+		delegation := types.MustUnmarshalDelegation(k.CDC(), iterator.Key(), iterator.Value())
+		if delegation.CrossStake {
+			ctx = ctx.WithCrossStake(true)
+		} else {
+			ctx = ctx.WithCrossStake(false)
+		}
+
+		result := handleMsgSideChainUndelegate(ctx, types.MsgSideChainUndelegate{
+			DelegatorAddr: delegation.DelegatorAddr,
+			ValidatorAddr: delegation.ValidatorAddr,
+			Amount:        sdk.NewCoin(boundDenom, delegation.GetShares().RawInt()),
+			SideChainId:   bscSideChainId,
+		}, k)
+		refundEvents = refundEvents.AppendEvents(result.Events)
+		if !result.IsOK() {
+			ctx.Logger().Debug("handleRefundStake failed",
+				"delegator", delegation.DelegatorAddr.String(),
+				"validator", delegation.ValidatorAddr.String(),
+				"amount", delegation.GetShares().String(),
+				"sideChainId", bscSideChainId,
+				"result", fmt.Sprintf("%+v", result),
+			)
+			// this is to prevent too many delegation is in unbounded state
+			// if too many delegation is in unbounded state, it will cause too many iteration in the block
+			failedCount++
+			if failedCount >= maxProcessedRefundFailed {
```

Comment:
At the edge case, when we reach `maxProcessedRefundFailed`, we cannot refund any records even in the following blocks.

Reply:
that is expected, if too many user undelegate by themself, we should wait for them to be completed.
but don't worry, undelegate is disable after secondSunsetFork, so the max time we wait for is 7days.